pypesto/optimize/optimize.py

```diff
@@ -25,71 +74,67 @@ def minimize(problem, optimizer,
+    if startpoint_method is None:
+        startpoint_method = uniform
 
-    """
+    # check options
+    if options is None:
+        options = OptimizeOptions()
+    options = OptimizeOptions.assert_instance(options)
 
-    # compute start points
-    if startpoint_method is False:
-        # fill with dummies
-        startpoints = np.zeros(n_starts, problem.dim)
-    else:
-        # apply startpoint method
-        startpoints = startpoint_method(n_starts,
-                                        problem.lb,
-                                        problem.ub,
-                                        problem.par_guesses)
-
-    # prepare result object
+    # assign startpoints
+    startpoints = assign_startpoints(n_starts, startpoint_method,
+                                     problem, options)
+
+    # prepare result
     if result is None:
         result = Result(problem)
 
     # do multistart optimization
-    for j in range(0, n_starts):
-        startpoint = startpoints[j, :]
-        if startpoint_resampling:
-            valid_startpoint = problem.objective(startpoint) < float('inf')
-            while not valid_startpoint:
-                startpoint = startpoint_method(
-                    1,
-                    problem.lb,
-                    problem.ub,
-                    problem.par_guesses
-                )[0, :]
-                valid_startpoint = problem.objective(startpoint) < float('inf')
+    for j_start in range(0, n_starts):
+        startpoint = startpoints[j_start, :]
 
+        # apply optimizer
         try:
-            optimizer_result = optimizer.minimize(problem, startpoint, j)
+            optimizer_result = optimizer.minimize(problem, startpoint, j_start)
         except Exception as err:
-            if allow_failed_starts:
-                print(('start ' + str(j) + ' failed: {0}').format(err))
-                optimizer_result = optimizer.recover_result(
-                    problem,
-                    startpoint,
-                    err
-                )
-            else:
-                raise
-
-        result.optimize_result.append(optimizer_result=optimizer_result)
+            optimizer_result = handle_exception(
+                options.allow_failed_starts,
+                problem.objective, startpoint, j_start, err)
+
+        # append to result
+        result.optimize_result.append(optimizer_result)
 
     # sort by best fval
     result.optimize_result.sort()
 
     return result
+
+
+def handle_exception(
+        allow_failed_starts,
+        objective, startpoint, j_start,
+        err) -> OptimizerResult:
+    """
+    Handle exceptions. Raise exception if allow_faile_starts is False,
+    otherwise return a dummy pypesto.OptimizerResult.
+    """
+    if allow_failed_starts:
+        print(('start ' + str(j_start) + ' failed: {0}').format(err))
+        optimizer_result = recover_result(objective, startpoint, err)
+        return optimizer_result
+    raise err
```

Comment:
does this properly rethrow the error with full traceback? If not, I would prefer the inline version.

Reply:
true, the stacktrace is wrong, starting in the new method. will change that.